### PR TITLE
Token exchange

### PR DESCRIPTION
**Pull request by IljaN**
## Related Issue

- Fixes <issue_link>

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] ...

### CODE REVIEW

File: lib/Application.php [hunk 1]
```diff
@@ -23,7 +23,7 @@
 
 require_once __DIR__ . '/../vendor/autoload.php';
 
-use Jumbojett\OpenIDConnectClientException;
```
Comment (DeepDiver1975) [inline lib/Application.php:26]: you can rebase to get rid of that ...

File: lib/Client.php [hunk 2]
```diff
@@ -90,6 +90,31 @@ public function __construct(
 		}
 	}
 
+	public function requestUserInfo($attribute = null) {
+		if (true) {
+			$resp = $this->requestTokenExchange(
+				$this->getRefreshToken(),
+				'urn:ietf:params:oauth:token-type:refresh_token',
```
Comment (DeepDiver1975) [inline lib/Client.php:97]: should be access token to have a solution which will work with our clients

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: lib/Client.php (hunk 2 above)
Comment (IljaN) [reply lib/Client.php:97]:
```
(
    [error_description] => Invalid request: Invalid token type for token exchange, must be a RefreshToken
    [error] => invalid_request
)
```

File: lib/Client.php [hunk 3]
```diff
@@ -90,6 +90,35 @@ public function __construct(
...
+				$this->getClientID()
+			);
+
+			$accessToken = $resp->access_token;
+			$userInfo = $this->introspectToken($accessToken);
```
Comment (DeepDiver1975) [inline lib/Client.php:102]: Just spotted this: this is anyhow wrong.

Request user info endpoint will only be for requesting the user info endpoint.

The logic to use either one or the other is handled in the app itself.

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: lib/Client.php (hunk 3 above)
Comment (IljaN) [reply lib/Client.php:102]: Where should this logic go to then? From what I gather requestUserInfo is always called. We could optimize this request away if we get the userinfo from the id-token....

Comment (IljaN) [reply lib/Client.php:102]: I guess this is a special case that /userinfo is disabled.

**Comment by IljaN** [comment]
Rebased with master

File: lib/Client.php [hunk 4]
```diff
@@ -127,9 +128,46 @@ public function getUserInfo() {
 			return $this->getAccessTokenPayload();
 		}
 
+		if (isset($openIdConfig['use-access-token-introspection-for-user-info']) && $openIdConfig['use-access-token-introspection-for-user-info']) {
+			$introspectionClientId = $openIdConfig['token-introspection-endpoint-client-id'] ?? null;
+			$introspectionClientSecret = $openIdConfig['token-introspection-endpoint-client-secret'] ?? null;
+			$accessToken = $this->getAccessToken();
```
Comment (DeepDiver1975) [inline lib/Client.php:134]: call not needed

File: lib/Client.php [hunk 5]
```diff
@@ -127,9 +128,46 @@ public function getUserInfo() {
...
+			if (isset($openIdConfig['exchange-token-mode-before-introspection'])) {
+				$mode = $openIdConfig['exchange-token-mode-before-introspection'];
+				$token = $mode === 'refresh-token' ? $this->getRefreshToken() : $this->getAccessToken();
```
Comment (DeepDiver1975) [inline lib/Client.php:137]: handle case where we have no refresh token -> throw exception/log error what ever.

This happens if oc is a rely party

File: lib/SessionVerifier.php [hunk 6]
```diff
@@ -103,6 +103,20 @@ public function verifySession(): void {
 			$introspectionClientId = $openIdConfig['token-introspection-endpoint-client-id'] ?? null;
 			$introspectionClientSecret = $openIdConfig['token-introspection-endpoint-client-secret'] ?? null;
 
+			if (isset($openIdConfig['exchange-token-mode-before-introspection'])) {
+				$mode = $openIdConfig['exchange-token-mode-before-introspection'];
+				$token = $mode === 'refresh-token' ? $this->session->get('oca.openid-connect.refresh-token') : $this->session->get('oca.openid-connect.access-token');
```
Comment (DeepDiver1975) [inline lib/SessionVerifier.php:108]: handle refresh token not being present - see above

File: lib/Client.php [hunk 7]
```diff
@@ -127,9 +128,46 @@ public function getUserInfo() {
...
+	 * @return string Access Token
+	 * @throws OpenIDConnectClientException
+	 */
+	public function exchangeToken(string $subjectToken, string $tokenType): string {
+		$subjectTokenType = $tokenType === 'refresh-token' ? 'urn:ietf:params:oauth:token-type:refresh_token' : 'urn:ietf:params:oauth:token-type:access_token';
+		$exchangeResponse = $this->requestTokenExchange($subjectToken, $subjectTokenType, $this->getClientID());
+
+		if (isset($exchangeResponse->error)) {
+			if (isset($exchangeResponse->error_description)) {
+				throw new OpenIDConnectClientException('TokenExchange response: ' . $exchangeResponse->error_description);
```
Comment (DeepDiver1975) [inline lib/Client.php:163]: error property should be part of the exception message as well

Comment (DeepDiver1975) [reply lib/Client.php:163]: but you will never get there - see https://github.com/JuliusPC/OpenID-Connect-PHP/blob/1a41cad55fb293590331758119296c2264b564d4/src/Client.php#L1124

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Comment by DeepDiver1975** [comment]
rebased version: https://github.com/owncloud/openidconnect/pull/236

**Pull request closed without merging**